go/types/blob.go

```diff
@@ -30,10 +29,99 @@ func NewEmptyBlob() Blob {
 	return Blob{newBlobLeafSequence(nil, []byte{}), &hash.Hash{}}
 }
 
-// BUG 155 - Should provide Write... Maybe even have Blob implement ReadWriteSeeker
+// ReaderAt interface. Eagerly loads requested byte-range from the blob p-tree.
+func (b Blob) ReadAt(p []byte, off int64) (n int, err error) {
+	// TODO: Support negative off?
+	d.PanicIfTrue(off < 0)
+
+	startIdx := uint64(off)
+	if startIdx >= b.Len() {
+		return 0, io.EOF
+	}
+
+	endIdx := startIdx + uint64(len(p))
+	if endIdx > b.Len() {
+		endIdx = b.Len()
+	}
+
+	if endIdx == b.Len() {
+		err = io.EOF
+	}
+
+	if startIdx == endIdx {
+		return
+	}
+
+	leaves, localStart := LoadLeafSequences(b, startIdx, endIdx)
+	endIdx = localStart + endIdx - startIdx
+	startIdx = localStart
+
+	for _, s := range leaves {
+		bl := s.(blobLeafSequence)
+
+		le := endIdx
+		ll := uint64(len(bl.data))
+		if le > ll {
+			le = ll
+		}
+		src := bl.data[startIdx:le]
+
+		copy(p[n:], src)
+		n += len(src)
+		endIdx -= le
+		startIdx = 0
+	}
+
+	return
+}
+
 func (b Blob) Reader() *BlobReader {
-	cursor := newCursorAtIndex(b.seq, 0, true)
-	return &BlobReader{b.seq, cursor, nil, 0}
+	return &BlobReader{b, 0}
+}
+
+func (b Blob) Copy(w io.Writer) (n int64) {
+	return b.CopyReadAhead(w, 1<<23 /* 8MB */, 6)
+}
+
+func (b Blob) CopyReadAhead(w io.Writer, chunkSize uint64, concurrency int) (n int64) {
+	bChan := make(chan chan []byte, concurrency)
```

Comment:
The outer chan is the mechanism by read-ahead is actually accomplished. The result is that the code tries to stay |concurrency| blocks of |chunkSize| ahead of the current consumption point.